File: 1604. Least Number of Unique Integers after K Removals/Solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    int findLeastNumOfUniqueInts(std::vector<int>& arr, int k) {
        std::unordered_map<int, int> freqMap;

        // Count the frequency of each element in the array
        for (int num : arr) {
            freqMap[num]++;
        }

        // Store frequencies in a vector
        std::vector<int> frequencies;
        for (const auto& entry : freqMap) {
            frequencies.push_back(entry.second);
        }

        // Sort frequencies in ascending order
        std::sort(frequencies.begin(), frequencies.end());

        // Iterate through frequencies and decrement k
        int uniqueCount = frequencies.size();
        for (int freq : frequencies) {
            if (k >= freq) {
                k -= freq;
                uniqueCount--;
            } else {
                break;
            }
        }

        return uniqueCount;
    }
};

File: 1604. Least Number of Unique Integers after K Removals/Solution.java
class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        
        HashMap< Integer ,Integer> map = new HashMap<>();

        for( int i:arr)  map.put( i , map.getOrDefault( i,0 )+1);

        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for( Map.Entry<Integer,Integer> i : map.entrySet() ) pq.add( i.getValue());
          
        while( k>0) { 

            int temp= pq.peek();
            k -= temp;
            if( k >= 0 ) pq.poll();
            
        }
        return pq.size();

           
    }
}